extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java

```diff
@@ -0,0 +1,146 @@
+        this.containsVersionValue = containsVersionValue;
+        this.versionFieldName = versionFieldName;
+
+        if (versionField != null) {
+            adjustVersionValue();
+        }
+    }
+
+    public static VersionHandler handle(Object entity, BsonDocument document) {
+        Optional<Field> versionFieldOptional = extractVersionField(entity);
+
+        if (versionFieldOptional.isEmpty()) {
+            return new VersionHandler(
+                    entity,
+                    buildUpdateQuery(null, document),
+                    null,
+                    false,
+                    false,
+                    null);
+        }
+
+        Field versionField = versionFieldOptional.get();
+        Long versionValue = extractVersionValue(versionField, entity);
+        String versionFieldName = versionField.getName();
+        BsonDocument query = buildUpdateQuery(versionFieldName, document);
+
+        return new VersionHandler(entity, query, versionField, true, versionValue != null, versionFieldName);
+    }
+
+    public boolean containsVersionAnnotationAndValue() {
+        return containsVersionAnnotation && containsVersionValue;
+    }
+
+    private static Optional<Field> extractVersionField(Object entity) {
+        Field[] fields = entity.getClass().getFields();
+        Field[] declaredFields = entity.getClass().getDeclaredFields();
+        return Stream.of(fields, declaredFields)
+                .flatMap(Stream::of)
+                .filter(field -> field.isAnnotationPresent(Version.class))
+                .findFirst();
+    }
+
+    private static Long extractVersionValue(Field versionField, Object entity) {
+        try {
+            if (versionField == null) {
+                return null;
+            }
+
+            versionField.setAccessible(true);
+            Long versionValue = (Long) versionField.get(entity);
+            return versionValue;
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error on set the version value");
+        }
+    }
+
+    private VersionHandler adjustVersionValue() {
+        try {
+            if (!containsVersionAnnotation) {
+                return this;
+            }
+
+            Long versionValue = (Long) versionField.get(entity);
+            versionField.set(entity, versionValue == null ? 0L : versionValue + 1);
+            return this;
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error on set the version value");
+        }
+    }
+
+    /**
+     * Utility method to check if the during an update there are documents not updated.
+     * Throws OptimisticLockException when found documents not updated.
+     */
+    public void checkHasNotAffectedResults(Long quantityResultAffected) {
+        if (containsVersionAnnotation && quantityResultAffected == 0) {
+            throwOptimisticLockException(entity);
+        }
+    }
```

Comment:
Duplicated with the same version inside the version handler class.

Reply:
adjusted.